### PR TITLE
Parallel router: improvements

### PR DESCRIPTION
**Pull request by duck2**
This branch has better partition tree building and can do it on every iteration. Also has debug logging to see the bottlenecks.

Full list of changes:
- Properly enforce `--num_workers`
- 3 new options for `run_vtr_task.py`
  - `-write_rr_graphs`: Write out respective rr graphs from each run in XML format
  - `-write_lookaheads`: Write out router lookaheads from each run
  - `-use_previous`: Use files from previous run. Can reuse `net`, `place`, `route`, `rr_graph` and `lookahead`.
    - example: `-use_previous run018:net,run018:place` for a routing-only run 
- Improvements to partition tree building: ParaDRo-like lookup, cutline no longer has a thickness
- `connections_inf.remaining_targets` is moved into `RouteTree` as a bitset
  - Uses less memory and keeps correct data most of the time
  - (This changes serial router behavior!)
- `RouteTree` got a write lock in preparation for #2367
- TBB enabled in all runners
- There are now 4 parallel test cases in `vtr_reg_strong`
- Google runners have `VPR_NUM_WORKERS` set to 4 (makes use of parallel tatum)

#### Types of changes

- [x] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by duck2** [comment]
Fails only in strong_odin, assert level=3 and sanitize=on. Oddly specific...

**Comment by duck2** [comment]
Can't reproduce the failure on wintermute with `ASSERT_LEVEL=3 SANITIZE=ON`. No idea why this would happen.

File: vpr/src/route/partition_tree.h [hunk 1]
```diff
@@ -64,3 +87,34 @@ class PartitionTree {
...
+        for (auto& line : lines) {
+            f << line << std::endl;
+        }
+        f.close();
+    }
+};
+#else
+class PartitionTreeDebug {
+  public:
+    static inline void log(std::string msg) { std::cout << msg << "\n"; }
```
Comment (duck2) [inline vpr/src/route/partition_tree.h:117]: fix when squashing

File: vpr/src/route/route_timing.cpp [hunk 2]
```diff
@@ -1364,14 +1349,13 @@ static void setup_routing_resources(int itry,
...
                                     CBRR& connections_inf,
-                                    std::vector<vtr::optional<const RouteTreeNode&>>& rt_node_of_sink,
                                     const t_router_opts& router_opts,
                                     bool ripup_high_fanout_nets) {
     /* Build and return a partial route tree from the legal connections from last iteration.
      * along the way do:
      * 	update pathfinder costs to be accurate to the partial route tree
-     * 	find and store the pins that still need to be reached in incremental_rerouting_resources.remaining_targets
-     * 	find and store the rt nodes that have been reached in incremental_rerouting_resources.reached_rt_sinks
+     * 	find and store the pins that still need to be reached in connections_inf.remaining_targets
```
Comment (duck2) [inline vpr/src/route/route_timing.cpp:1357]: fix when squashing (no such thing as connections_inf.remaining_targets)

File: vtr_flow/scripts/run_vtr_flow.py [hunk 3]
```diff
@@ -583,7 +590,7 @@ def vtr_command_main(arg_list, prog=None):
         )
 
     except KeyboardInterrupt as error:
-        print("{} recieved keyboard interrupt".format(prog))
+        print("{}  keyboard interrupt".format(prog))
```
Comment (duck2) [inline vtr_flow/scripts/run_vtr_flow.py:593]: fix when squashing

**Comment by duck2** [comment]
Should be ready for review @vaughnbetz

**Comment by duck2** [comment]
Here's a single threaded QoR result for `titan_quick_qor`, seed=10:

**metric**|**master**|**par_router2**
:-----:|:-----:|:-----:
vtr\_flow\_elapsed\_time|1|0,9996106575
max\_vpr\_mem|1|1,004752084
routed\_wirelength|1|1,000248005
critical\_path\_delay|1|1,004835426
geomean\_nonvirtual\_intradomain\_critical\_path\_delay|1|1,008395633
crit\_path\_route\_time|1|1,006238626

**Comment by duck2** [comment]
@saaramahmoudi

**Comment by soheilshahrouz** [comment]
The NoC test bug was fixed in 7feda705. Feel free to cherrypick the commit.
The problem was that the randomly generated NocRouterId was out of range.

**Comment by vaughnbetz** [comment]
Excellent, thanks @soheilshahrouz ! 
@duck2 : can you rebase or cherry pick that, update the (one?) golden result that's failing and then merge this (or I can merge it if you prefer).

**Pull request merged**